File: Android_app/pioneer/src/main/java/com/ABC/pioneer/sensor/payload/Crypto/PioneerHMac.java
package com.ABC.pioneer.sensor.payload.Crypto;

public class PioneerHMac {
    Digest digest;
    Mac hmac;

    public PioneerHMac(byte[] key)
    {
        digest = new SM3Digest();
        hmac = new HMac(digest);
        hmac.init(new KeyParameter(key));
    }

    public void generateHMac(byte[] message,byte[] mac)
    {
        generateHMac(message,0,message.length,mac,0);
    }

    public void generateHMac(byte[] message,int inoff,int len,byte[] mac,int outoff)
    {
        hmac.update(message,inoff,len);
        hmac.doFinal(mac,outoff);
    }
}

File: IOS_app/Pioneer_ios/Pioneer Framework/Pioneer/Crypto/PioneerHMac.swift
//
//  PioneerHMac.swift
//  Pioneer
//
//  Created by Beh on 2021/6/3.
//

import Foundation

public class PioneerHMac
{
    var digest: Digest
    var hmac: MAC
    
    init(from key: [UInt8])
    {
        digest = SM3Digest()
        hmac = HMAC(which:digest)
        hmac.macinit(from: KeyParameter(from:key))
    }
    
    public func generateMac(from message: [UInt8],to mac: inout [UInt8])
    {
        generateMac(from: message, inOff: 0, len: message.count, to: &mac, outOff: 0)
    }
    
    public func generateMac(from message: [UInt8], inOff: Int, len: Int, to mac: inout [UInt8], outOff: Int)
    {
        hmac.update(from: message, inOff: inOff, len: len)
        _ = hmac.doFinal(to: &mac, outOff: outOff)
    }
}
